peps/pep-0728.rst

```diff
@@ -632,99 +628,112 @@ For example::
 :mod:`typing` documentation.
 
 Backwards Compatibility
 =======================
 
-Because ``__extra_items__`` remains as a regular key if ``closed=True`` is not
-specified, no existing codebase will break due to this change.
-
-If the proposal is accepted, none of ``__required_keys__``,
-``__optional_keys__``, ``__readonly_keys__`` and ``__mutable_keys__`` should
-include ``"__extra_items__"`` defined on the same TypedDict type when
-``closed=True`` is specified.
+Because ``extra_items`` is an opt-in feature, no existing codebase will break
+due to this change.
 
-Note that ``closed`` as a keyword argument does not collide with the keyword
-arguments alternative to define keys with the functional syntax that allows
-things like ``TD = TypedDict("TD", foo=str, bar=int)``, because it is scheduled
-to be removed in Python 3.13.
+Note that ``closed`` and ``extra_items`` as keyword arguments do not collide
+with othere keys when using something like
+``TD = TypedDict("TD", foo=str, bar=int)``, because this syntax is already
+scheduled to be removed in Python 3.13.
 
 Because this is a type-checking feature, it can be made available to older
 versions as long as the type checker supports it.
 
+Open Issues
+===========
+
+Use a Special ``__extra_items__`` Key with the ``closed`` Class Parameter
+-------------------------------------------------------------------------
+
+In an earlier revision of this proposal, we discussed an approach that would
+utilize ``__extra_items__``'s value type to specify the type of extra items
+accepted, like so::
+
+    class IntDict(TypedDict, closed=True):
+        __extra_items__: int
+
+where ``closed=True`` is required for ``__extra_items__`` to be treated
+specially, to avoid key collision.
+
+Some members of the community concern about the elegance of the syntax.
+Practiaclly, the key collision with a regular key can be mitigated with
+workarounds, but since using a reserved key is central to this proposal,
+there are limited ways forward to address the concerns.
+
+Support a New Syntax of Specifying Keys
+---------------------------------------
+
+By introducing a new syntax that allows specifying string keys, we could
+deprecate the functional syntax of defining TypedDict types and address the
+key conflict issues if we decide to reserve a special key to type extra items.
+
+For example::
+
+    class Foo(TypedDict):
+        name: str  # Regular item
+        _: bool    # Type of extra items
+        __items__ = {
+            "_": int,   # Literal "_" as a key
+            "class": str,  # Keyword as a key
+            "tricky.name?": float,  # Arbitrary str key
+        }
+
+This was proposed `here by Jukka
+<https://discuss.python.org/t/pep-728-typeddict-with-typed-extra-items/45443/115>`__.
+
+This will allow us to deprecate the functional syntax of defining TypedDict
+types altogether, but adapting to this proposal will make this PEP heavier.
```

Comment:
I feel this isn't a strong argument; if this PEP is accepted, we'll be stuck with its behavior for at least many years, so we need to make sure we get it right the first time.

This proposal is nice because it also unlocks some other things that are currently awkward or impossible (keys that aren't valid identifiers), it has a few disadvantages:
- It's less apparent to a reader that `_: bool` makes the TypedDict special, relative to adding a class argument like `extra_items=bool`.
- It's backwards incompatible with existing TypedDicts using the `_: bool` key. While such users have a way to get around the issue, it's still a problem for them if they upgrade Python (or typing-extensions).
- The types don't appear in an annotation context, so their evaluation will not be deferred.

Reply:
Added the bullet points to this section in commit d146b98208925db3ce8ecf683b5c81dbc7d274c5. Thanks!